File: Page Replacement java/lru.java
import java.util.*;

public class lru {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the reference string: ");
        String refString = sc.nextLine();
        System.out.print("Enter the memory size: ");
        int memSize = sc.nextInt();

        int pageFaults = 0;
        int pageHits = 0;
        List<Character> memory = new ArrayList<>();
        LinkedHashMap<Character, Integer> pageMap = new LinkedHashMap<>();

        System.out.println("\nPage replacement process:");
        for (char page : refString.toCharArray()) {
            System.out.println("Page: " + page);
            if (memory.contains(page)) {
                pageHits++;
                pageMap.remove(page);
                pageMap.put(page, pageMap.size());
            } else {
                pageFaults++;
                if (memory.size() < memSize) {
                    memory.add(page);
                    pageMap.put(page, pageMap.size());
                } else {
                    char leastRecentlyUsed = pageMap.entrySet().iterator().next().getKey();
                    pageMap.remove(leastRecentlyUsed);
                    memory.remove(Character.valueOf(leastRecentlyUsed));
                    memory.add(page);
                    pageMap.put(page, pageMap.size());
                }
            }
            printMemoryState(memory);
        }

        double pageFaultRate = (double) pageFaults / (pageFaults + pageHits);
        double pageHitRate = (double) pageHits / (pageFaults + pageHits);

        System.out.println("\nNumber of page faults: " + pageFaults);
        System.out.println("Number of page hits: " + pageHits);
        System.out.println("Page fault rate: " + pageFaultRate);
        System.out.println("Page hit rate: " + pageHitRate);
    }

    private static void printMemoryState(List<Character> memory) {
        System.out.print("Memory state: ");
        for (char page : memory) {
            System.out.print(page + " ");
        }
        System.out.println();
    }
}

File: c cpp soham/Memory Management/lru.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    cout << "Enter the reference string (without space): ";
    string refString;
    cin >> refString;

    cout << "Enter the Number of frames: ";
    int memSize;
    cin >> memSize;

    int pageFaults = 0;
    int pageHits = 0;
    list<char> memory;
    unordered_map<char, list<char>::iterator> pageMap;

    cout << "\nPage replacement process:" << endl;
    for (char page : refString)
    {
        cout << "Page: " << page << endl;
        if (pageMap.find(page) != pageMap.end())
        {
            // Page Hit
            pageHits++;
            // Move the accessed page to the front (most recently used)
            memory.erase(pageMap[page]);
            memory.push_front(page);
            pageMap[page] = memory.begin();
        }
        else
        {
            // Page Fault
            pageFaults++;
            if (memory.size() < memSize)
            {
                // If memory is not full, simply add the page
                memory.push_front(page);
                pageMap[page] = memory.begin();
            }
            else
            {
                // If memory is full, remove the least recently used page
                char removedPage = memory.back();
                memory.pop_back();
                pageMap.erase(removedPage);
                // Add the new page to the front
                memory.push_front(page);
                pageMap[page] = memory.begin();
            }
        }
        cout << "Memory state: ";
        for (char p : memory)
        {
            cout << p << " ";
        }
        cout << endl;
    }

    double pageFaultRate = static_cast<double>(pageFaults) / (pageFaults + pageHits);
    double pageHitRate = static_cast<double>(pageHits) / (pageFaults + pageHits);

    cout << "\nNumber of page faults: " << pageFaults << endl;
    cout << "Number of page hits: " << pageHits << endl;
    cout << "Page fault rate: " << pageFaultRate << endl;
    cout << "Page hit rate: " << pageHitRate << endl;

    return 0;
}
